go/vt/orchestrator/logic/topology_recovery.go

```diff
@@ -1305,68 +1210,61 @@ func checkAndExecuteFailureDetectionProcesses(analysisEntry inst.ReplicationAnal
 	return true, true, err
 }
 
+// getCheckAndRecoverFunction gets the recoveryFunction to use for the given analysis.
+// It also returns an integer which is supposed to be unique for each function that we return.
+// It is used for checking the equality of the returned function.
 func getCheckAndRecoverFunction(analysisCode inst.AnalysisCode, analyzedInstanceKey *inst.InstanceKey) (
-	checkAndRecoverFunction func(analysisEntry inst.ReplicationAnalysis, candidateInstanceKey *inst.InstanceKey, forceInstanceRecovery bool, skipProcesses bool) (recoveryAttempted bool, topologyRecovery *TopologyRecovery, err error),
+	checkAndRecoverFunction func(ctx context.Context, analysisEntry inst.ReplicationAnalysis, candidateInstanceKey *inst.InstanceKey, forceInstanceRecovery bool, skipProcesses bool) (recoveryAttempted bool, topologyRecovery *TopologyRecovery, err error),
+	recoverFunctionCode int,
 	isActionableRecovery bool,
 ) {
 	switch analysisCode {
 	// primary
 	case inst.DeadPrimary, inst.DeadPrimaryAndSomeReplicas:
 		if isInEmergencyOperationGracefulPeriod(analyzedInstanceKey) {
-			return checkAndRecoverGenericProblem, false
+			return checkAndRecoverGenericProblem, 1, false
 		}
-		return checkAndRecoverDeadPrimary, true
+		return recoverDeadPrimary, 2, true
```

Comment:
I am a little unsure of this way. We need some way to compare that the function we get back are the same, but golang doesn't support equality comparison on functions. So, I have returned a integer for each of the functions, which is unique for it and used it for checkin equality. If we prefer, I can make this an enum and give the integers more meaningful variable names. Maybe @vmg has a better solution?

Reply:
You can register the function and then check if they correspond to the same.

Reply:
Right, so a map from a string to the function and use the string for checking equality?

Reply:
You could simply return the same of the function as a string, e.g. `"recoverDeadPrimary"` and compare that. You don't need a map. I think pretty much anything works here, the only difficulty is making sure that you're not reusing IDs, which will lead to bugs, which is why returning a number seems questionable.

Reply:
Thankyou @vmg That makes sense. I was feeling iffy with the integers too. I'll change them to an enum instead, which would give them more meaningful names and we would also not have the problems of typing errors in the name of the returned function!

Reply:
Fixed it.